docs/tutorials/essentials/part-8-rtk-query-advanced.md

```diff
@@ -0,0 +1,1074 @@
+
+RTK Query deliberately **does _not_ implement a cache that would deduplicate identical items across multiple requests**. There are several reasons for this:
+
+- A fully normalized shared-across-queries cache is a _hard_ problem to solve
+- We don't have the time, resources, or interest in trying to solve that right now
+- In many cases, simply re-fetching data when it's invalidated works well and is easier to understand
+- At a minimum, RTKQ can help solve the general use case of "fetch some data", which is a big pain point for a lot of people
+
+In comparison, we just normalized the response data for the `getUsers` endpoint, in that it's being stored as an `{id: value}` lookup table. However, **this is _not_ the same thing as a "normalized cache" - we only transformed _how this one response is stored_** rather than deduplicating results across endpoints or requests.
+
+### Selecting Values from Results
+
+The last component that is reading from the old `postsSlice` is `<UserPage>`, which filters the list of posts based on the current user. We've already seen that we can get the entire list of posts with `useGetPostsQuery()` and then transform it in the component, such as sorting inside of a `useMemo`. The query hooks also give us the ability to select pieces of the cached state by providing a `selectFromResponse` option, and only re-render when the selected pieces change.
+
+We can use `selectFromResponse` to have `<UserPage>` read just a filtered list of posts from the cache. However, in order for `selectFromResponse` to avoid unnecessary re-renders, we need to ensure that whatever data we extract is memoized correctly. To do this, we should create a new selector instance that the `<UsersPage>` component can reuse every time it renders, so that the selector memoizes the result based on its inputs.
+
+```jsx title="features/users/UsersPage.js"
+// highlight-next-line
+import { createSelector } from '@reduxjs/toolkit'
+
+import { selectUserById } from '../users/usersSlice'
+// highlight-next-line
+import { useGetPostsQuery } from '../api/apiSlice'
+
+export const UserPage = ({ match }) => {
+  const { userId } = match.params
+
+  const user = useSelector(state => selectUserById(state, userId))
+
+  // highlight-start
+  const selectPostsForUser = useMemo(() => {
+    // Return a unique selector instance for this page so that
+    // the filtered results are correctly memoized
+    return createSelector(
+      res => res.data,
+      (res, userId) => userId,
+      (data, userId) => data.filter(post => post.user === userId)
+    )
+  }, [])
+  // highlight-end
+
+  // Use the same posts query, but extract only part of its data
+  const { postsForUser } = useGetPostsQuery(undefined, {
+    // highlight-start
+    selectFromResult: result => ({
+      ...result,
+      // Include a field called `postsForUser` in the hook result object,
+      // which will be filtered list of posts
+      postsForUser: selectPostsForUser(result, userId)
+    })
```

Comment:
Looking at the source, pretty sure it runs into this `shallowEqual` check here:

```ts
      const currentState = useSelector(
        (state: RootState<Definitions, any, any>) =>
          querySelector(state, lastValue.current),
        shallowEqual
      )
```

so spread + add field with a memoized selector should work, I think.

Reply:
It definitely works either way, but what I mean is that it will better optimized for renders if you exclude the rest of `result`, because including it means that more things are being checked in the `shallowEqual` comparison, which in this case isn't being used.

So for this case:
```ts
    selectFromResult: result => ({
      ...result,
      postsForUser: selectPostsForUser(result, userId)
    })
```
If *any part* of result changes, the hook will re-render. i.e. if posts for other users update, or even when changing status between `pending` & `fulfilled`.

Whereas when removing the spread `...result` like this case:
```ts
    selectFromResult: result => ({
      postsForUser: selectPostsForUser(result, userId)
    })
```
This will only re-render specifically if `postsForUser` changes, without caring about the rest of the values in `...result`. Since `postsForUser` is the *only* thing being used by the caller, this is a strictly more optimized version with no downside

Reply:
FYI here's a contrived example that demonstrates the behaviour: https://codesandbox.io/s/rtk-query-fiddlebox-selectfromresult-dx17l?file=/src/components/TodoApp.tsx

- Two components, one with each scenario above:
  - Using a list query, one spreading the entire result, one without, both with an added field
- The added field picks out a particular item (todo with ID 2)
- Each logs something in the render body
- Use the UI to update the message for a different todo (with ID 1) - i.e. part of the result that the hook caller is not using
- Observe that the version with the spread result will re-render while the one without does not re-render

Reply:
Ok, I'll add a note along those lines.